### PR TITLE
SSH2Stream: ssh2-streams throws on receiving multiple KEXDH_INIT

### PR DESCRIPTION
**Pull request by nrjpoddar**
packet

ssh2-streams performs verification on receiving the packet that the
packet matches the expected packet. This check
however doesn't work for KEXDH_* packets as the context specific message codes
are not added to constants.Message object. As the packet types are missing in
constants.Message, packet verification in parsePacket doesn't work for KEXDH_* packets
subsequently creating an exception. Added code which verifies that the received
KEXDH_* packets is the next expected packet.

There was a second bug where after receiving NEWKEYS which causes
outstate.status to be OUT_READY, if KEXDH_* packet are received it also
created an exception. The check in parsePacket is skipped if the state is
OUT_READY and KEXDH_* packets are parsed and passed up the chain even though
this is an invalid packet. Added check to enusre if KEXDH_* packets are received
after OUT_READY session is disconnected.

Added unit tests which verifies the behavior of receiving
multiple KEXDH_INIT messages.

Server side code doesn't exist for dynamic group KEXDH algorithms hence
multiple KEXDH_GEX_* packets cannot be currently verified.

### CODE REVIEW

**Comment by nrjpoddar** [comment]
Brian,

I have another pull request which fixes protocol errors and has a unit test associated with it. In order for unit test to pass I need _send that I have exposed in this pull request. If you can take a look at this request and provide feedback/merge I can submit subsequent pull requests. Thanks!

File: lib/ssh.js [hunk 1]
```diff
@@ -219,6 +221,16 @@ SSH2Stream.prototype.push = function(chunk, encoding) {
   return ret;
 };
 
+// This is only used to send raw packets from unit tests
```
Comment (mscdex) [inline lib/ssh.js]: I'd probably rather just export `send()` as a "private" method (e.g. `_send()` or similar).

File: lib/ssh.js [hunk 2]
```diff
@@ -3831,8 +3843,20 @@ function parse_KEXINIT(self, callback) {
 function parse_KEX(self, type, callback) {
   var state = self._state,
       instate = state.incoming,
-      payload = instate.payload;
+      outstate = state.outgoing,
+      payload = instate.payload,
+      pktType = (state.kexdh !== 'group') ? DYNAMIC_KEXDH_MESSAGE[type] :
```
Comment (mscdex) [inline lib/ssh.js]: Style nit: If a ternary has to be split, it should look like:

```javascript
(state.kexdh !== 'group
 ? DYNAMIC_KEXDH_MESSAGE[type]
 : KEXDH_MESSAGE[type]);
```

File: lib/ssh.js [hunk 3]
```diff
@@ -3831,8 +3843,20 @@ function parse_KEXINIT(self, callback) {
...
+                                            KEXDH_MESSAGE[type];
 
+  if ((outstate.status === OUT_READY) ||
```
Comment (mscdex) [inline lib/ssh.js]: Style nit: If complex conditionals need to be split across new lines, the operator should be first on the next line. Also, extra parens should be removed:

```javascript
if (outstate.status === OUT_READY
    || instate.expectedPacket !== pktType) {
```

File: test/test-kexdh.js [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+          outstate = state.outgoing,
+          buf = new Buffer(1 + 4 + outstate.pubkey.length);
+      buf[0] = MESSAGE.KEXDH_INIT;
+      buf.writeUInt32BE(outstate.pubkey.length, 1, true);
+      outstate.pubkey.copy(buf, 5);
+      client._send(buf, undefined, true);
+    });
+    client.on('error', function (err) {
+      assert.equal('PROTOCOL_ERROR', err.message,
+        makeMsg(what, 'Expected Error: PROTOCOL_ERROR Got Error: ' +
```
Comment (mscdex) [inline test/test-kexdh.js]: Style nit: the operator should move to the start of the next line. Also, split all or no function arguments for consistency/easier readability, like:

```javascript
assert.equal('PROTOCOL_ERROR',
                     err.message,
                     makeMsg(what,
                                      'Expected PROTOCOL_ERROR Got: '
                                        + err.message));
```

File: test/test-kexdh.js [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    path = require('path'),
+    fs = require('fs');
+
+var t = -1,
+    group = path.basename(__filename, '.js') + '/',
+    SERVER_KEY = fs.readFileSync(__dirname + '/fixtures/ssh_host_rsa_key');
+
+function multi_kexdh_init() {
+  function run() {
+    var what = 'MULTI_KEXDH_INIT: ';
```
Comment (mscdex) [inline test/test-kexdh.js]: Minor nit: This could probably be simplified more to just be something like: `'Multi-KEXDH init'`

**Comment by mscdex** [comment]
A few nits, but otherwise LGTM.

**Comment by mscdex** [comment]
Landed in 9af07ba6e7ce5d522a7595380cf5dfec7eeb0a93 with a few changes and updates. Thanks!

**Pull request closed without merging**